libraries/AP_BattMonitor/AP_BattMonitor_TIBQ76952.cpp

```diff
@@ -0,0 +1,937 @@
+
+    // Mark configuration as complete to prevent repeated attempts
+    configured = true;
+    Debug("BQ76952: Configuration complete");
+
+    // report success
+    return true;
+}
+
+// read battery voltage, current and temperature
+void AP_BattMonitor_TIBQ76952::read_voltage_current_temperature()
+{
+    // exit immediately if full voltage scan has not completed (FULL_SCAN bit 7)
+    uint16_t alarm_raw_status = direct_command_read_2bytes(TIBQ769x2_AlarmRawStatus);
+    if (!BIT_IS_SET(alarm_raw_status, 7)) {
+        return;
+    }
+
+    // get semaphore before updating accumulate structure
+    WITH_SEMAPHORE(accumulate_sem);
+
+    // ready stack voltage (should equal sum of cell voltages)
+    // we do not use the package voltage because this is floating when FETs are off
+    accumulate.voltage += direct_command_read_2bytes(TIBQ769x2_StackVoltage) * 0.01;
+
+    // read individual cell voltages
+    for (uint8_t i = 0; i < HAL_BATTMON_CELL_COUNT; i++) {
+        uint16_t cell_voltage_mv = direct_command_read_2bytes(TIBQ769x2_Cell1Voltage + i*2);
+        accumulate.cell_voltages_mv[i] += cell_voltage_mv;
+    }
+
+    // read current (positive values = discharging, negative = charging)
+    int16_t cc2_current = direct_command_read_2bytes(TIBQ769x2_CC2Current);
+    accumulate.current += cc2_current * 0.001f; // convert to Amps
+
+    // read temperature
+    int16_t temp_internal = direct_command_read_2bytes(TIBQ769x2_IntTemperature); // 0.1K
+    accumulate.temp += KELVIN_TO_C(temp_internal * 0.1f); // convert to degC
+
+    // increment number of readings
+    accumulate.count++;
+}
+
+// ready battert power state (e.g. charging, discharging, idle)
+void AP_BattMonitor_TIBQ76952::read_power_state()
+{
+    AP_BattMonitor::PowerState new_state = _state.power_state;
+
+    uint16_t mfg_status = direct_command_read_2bytes(TIBQ769x2_MfgStatusInit);
+    uint8_t fet_status = direct_command_read_1byte(TIBQ769x2_FETStatus);
+
+    if (mfg_status & MfgStatusInit_FET_EN) {
+        new_state = AP_BattMonitor::PowerState::IDLE;
+        if (fet_status & FET_STATUS_CHG_FET_EN && accumulate.current > 0) {
+            // Charging if CHG_FET bit is set
+            new_state = AP_BattMonitor::PowerState::CHARGING;
+        }
+        if (fet_status & FET_STATUS_DSG_FET_EN) {
+            // Discharging if DSG_FET bit is set
+            new_state = AP_BattMonitor::PowerState::DISCHARGING;
+        }
+    }
+
+    _state.power_state = new_state;
+}
+
+// ready safety status (e.g. check for faults)
+void AP_BattMonitor_TIBQ76952::read_safety_state()
+{
+    uint16_t safetyA_status = direct_command_read_2bytes(TIBQ769x2_SafetyStatusA);
+    if (safetyA_status > 0) {
+        Debug("BQ76952: SafetyA Status: 0x%04X", safetyA_status);
+    }
+}
+
+// read bytes from a register. returns true on success
+bool AP_BattMonitor_TIBQ76952::read_register(uint8_t reg_addr, uint8_t *reg_data, uint8_t len) const
+{
+    WITH_SEMAPHORE(dev->get_semaphore());
```

Comment:
Have you tested this driver when `dev` ends up as nullptr?  (actually just `!dev` when using ownptr)